protected/humhub/components/bootstrap/ModuleAutoLoader.php

```diff
@@ -43,11 +62,52 @@ public function bootstrap($app)
                     }
                 }
             }
-            if (!YII_DEBUG) {
-                Yii::$app->cache->set(self::CACHE_ID, $modules);
+            Yii::$app->cache->set(self::CACHE_ID, $modules);
+        }
+
+        return $modules;
+    }
+
+    /**
+     * Find all modules with configured paths
+     * @param array $paths
+     * @return array
+     */
+    public static function findModules($paths)
+    {
+        $folders = [];
+        foreach ($paths as $path) {
+            $folders = array_merge($folders, self::findModulesByPath($path));
+        }
+
+        $modules = [];
+        foreach ($folders as $folder) {
+            try {
+                /** @noinspection PhpIncludeInspection */
+                $modules[$folder] = require $folder . DIRECTORY_SEPARATOR . self::CONFIGURATION_FILE;
+            } catch (\Exception $e) {
+                Yii::error($e);
             }
         }
 
-        Yii::$app->moduleManager->registerBulk($modules);
+        return $modules;
+    }
+
+    /**
+     * Find all directories with a configuration file inside
+     * @param string $path
+     * @return array
+     */
+    public static function findModulesByPath($path)
+    {
+        $hasConfigurationFile = function ($path) {
+            return is_file($path . DIRECTORY_SEPARATOR . self::CONFIGURATION_FILE);
+        };
+
+        try {
+            return FileHelper::findDirectories(Yii::getAlias($path, true), ['filter' => $hasConfigurationFile]);
```

Comment:
https://www.yiiframework.com/doc/api/2.0/yii-helpers-basefilehelper#findDirectories()-detail

What about the case when

- moduleDir
   - tests
      - config.php
   - config.php

Reply:
Good catch. I will look into recursive true/false tomorrow.